proposals/0016-c-interface-compiler-library.md

```diff
@@ -50,7 +50,495 @@ older DXC-based solutions to adopt clang as the preferred HLSL compiler.
 
 ## Detailed design
 
-_The detailed design is not required until the feature is under review._
+### What is provided today in the DXC api?
+The current DirectX shader compiler library is a nano-COM implementation that
+supports the following features.
+
+* [Include Handlers](#include-handlers)
+* [Compiler](#shader-compiler)
+* [Linker](#shader-linker)
+* [Validation](#shader-validation)
+* [Reflection Data Access](#shader-reflection)
+* [DXIL Container Access](#dxil-container-access)
+* [PDB Symbol Access](#shader-pdbs)
+
+These features will also need to be supported by the new c-style library.
+Code snippets on how to compile a shader using the COM library are located
+in [DXC api code snippets](#dxc-code-snippets).
+
+
+The following interfaces are used to work with data being passed to/from the
+library.
+
+#### Buffers
+```c++
+// Structure for supplying bytes or text input to Dxc APIs. Represents both
+// text (DxcText) and non-text byte buffers (DxcBuffer).
+struct DxcBuffer {
+  LPCVOID Ptr;
+  SIZE_T Size;
+  UINT Encoding;
+};
+typedef DxcBuffer DxcText;
+
+// General purpose buffers
+struct IDxcBlob : public IUnknown {
+public:
+  LPVOID GetBufferPointer();
+  SIZE_T GetBufferSize();
+};
+
+// String buffers that guarantee null-terminated text and the stated encoding
+struct IDxcBlobEncoding : public IDxcBlob {
+public:
+  HRESULT GetEncoding(BOOL *pKnown, UINT32 *pCodePage);
+};
+
+struct IDxcBlobUtf16 : public IDxcBlobEncoding {
```

Comment:
Are these proper interfaces with uuid's that have been omitted from this spec, so that you can QI for IDxcBlobUtf16?

Reply:
I removed the uuids, but yes.  IDXcBlob is built on IUnknown. and Encoding is from Blob, and Utf16 is from Encoding. . I am not sure QI works all the way through as I have not looked at the interface impl.  If COM rules apply then I would think it would be possible.. But nano-COM sometimes changes things.

Reply:
If QI doesn't work as you'd expect it to then it isn't nano-COM.